File: src/main/java/rbq/wtf/lycoris/client/wrapper/wrappers/wrapper/IWrapper.java
package rbq.wtf.lycoris.client.wrapper.wrappers.wrapper;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.Objects;

public class IWrapper {
    private final Object object;

    public IWrapper(Object obj) {
        this.object = obj;
    }

    public Object getWrapObject() {
        return object;
    }

    protected Object getField(Field targetField) {
        try {
            return targetField.get(getWrapObject());
        } catch (Exception e) {
            e.printStackTrace();
            throw new NullPointerException("Can't get Field: " + e.getMessage());
        }
    }

    protected void setField(Field targetField, Object value) {
        try {
            value = value instanceof IWrapper ?
                    ((IWrapper) value).getWrapObject() : value;
            targetField.set(getWrapObject(), value);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    protected Object invoke(Method targetMethod, Object... args) {
        try {
            for (int i = 0; i < args.length; i++) {
                if (args[i] instanceof IWrapper)
                    args[i] = ((IWrapper) args[i]).getWrapObject();
            }
            return targetMethod.invoke(getWrapObject(), args);
        } catch (Exception e) {
            e.printStackTrace();
            throw new NullPointerException("Can't get Invoke Method: " + e.getMessage());
        }
    }

    public boolean isNull() {
        return Objects.isNull(object);
    }

    public boolean equals(IWrapper wrapper) {
        return Objects.equals(object, wrapper.object);
    }
}

File: src/main/java/rbq/wtf/lycoris/client/wrapper/wrappers/wrapper/IWrapper.kt
package rbq.wtf.lycoris.client.wrapper.wrappers.wrapper

import rbq.wtf.lycoris.client.utils.Logger
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.util.*

open class IWrapper(val wrapObject: Any) {

    protected fun getField(targetField: Field): Any? {
        return try {
            targetField.get(wrapObject)
        } catch (e: Exception) {
            e.printStackTrace()
            throw NullPointerException("Can't get Field: " + e.message)
        }
    }

    protected fun setField(targetField: Field, value: Any?) {
        try {
            targetField[wrapObject] = if (value is IWrapper) value.wrapObject else value
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    protected operator fun invoke(targetMethod: Method, vararg args: Any?): Any? {
        return try {
            val varargs: Array<Any?> = arrayOf(*args)
            for (i in varargs.indices) {
                if (varargs[i] is IWrapper) varargs[i] = (varargs[i] as IWrapper).wrapObject
            }
            targetMethod.invoke(wrapObject, *varargs)
        } catch (e: Exception) {
            e.printStackTrace()
            throw NullPointerException("Can't get Invoke Method: " + e.message)
        }
    }

    val isNull: Boolean
        get() = Objects.isNull(wrapObject)

    override fun equals(other: Any?): Boolean { //只会判断WrapObject
        return if (other is IWrapper) wrapObject == other.wrapObject else wrapObject == other
    }

    override fun hashCode(): Int {
        return wrapObject.hashCode()
    }
}